# Grace's First Notebook

## Working on getting to know the data, thinking of some research questions

### Possible questions:

- How much are top merchants spending per order? Are we retaining these top buyers?
- On average, what are me making per order? 
- Do we earn more from bulk purchases from large merchants, or from smaller, more consistent buyers?
- What are the top items purchased? Has interest increased over time?
- Who are our most consistent customers? (Consistent = buying significant product regularly over a long span of time)

## Business questions:
- Is it possible to play multiple orders at the same time? In previous notebook, some of the different orders placed at the same millisecond were by the same merchant. 

In [1]:
import pandas as pd
import numpy as np

In [2]:
baskets = pd.read_csv('baskets_sample_random_10.csv')

## Simple tests to figure out how to read data

In [4]:
baskets.count()

id             29305
order_id       29305
placed_at      29305
merchant_id    29305
sku_id         29305
top_cat_id     29298
sub_cat_id     29298
qty            29305
price          29305
dtype: int64

In [5]:
baskets[baskets['price'] > 292000.0].count()

id             2509
order_id       2509
placed_at      2509
merchant_id    2509
sku_id         2509
top_cat_id     2509
sub_cat_id     2509
qty            2509
price          2509
dtype: int64

In [6]:
baskets[baskets['sku_id'] == 702].count()

id             218
order_id       218
placed_at      218
merchant_id    218
sku_id         218
top_cat_id     218
sub_cat_id     218
qty            218
price          218
dtype: int64

### Snapshot of dataframe 

In [8]:
baskets[1:]

id  order_id                placed_at  merchant_id  sku_id  \
1         166        41  2021-05-06 10:45:02.448          196     341   
2         167        42  2021-05-06 10:45:04.850          196     341   
3         168        43  2021-05-06 10:48:00.927          197     457   
4         169        44  2021-05-06 10:51:28.980          198     341   
5         170        44  2021-05-06 10:51:28.980          198     457   
...       ...       ...                      ...          ...     ...   
29300  338870     61206  2022-07-31 16:52:33.731          428    1896   
29301  338871     61206  2022-07-31 16:52:33.731          428     747   
29302  338872     61206  2022-07-31 16:52:33.731          428     850   
29303  338873     61206  2022-07-31 16:52:33.731          428     853   
29304  338874     61206  2022-07-31 16:52:33.731          428     852   

       top_cat_id  sub_cat_id  qty     price  
1             3.0        47.0    2       0.0  
2             3.0        47.0    2       0.0  
3            12.0        30.0   20       0.0  
4             3.0        47.0    5       0.0  
5            12.0        30.0   10       0.0  
...           ...         ...  ...       ...  
29300         4.0        57.0   80   12350.0  
29301        12.0        36.0    1   86000.0  
29302         9.0        48.0    1  101500.0  
29303         9.0        48.0    1   68500.0  
29304         9.0        48.0    1   68500.0  

[29304 rows x 9 columns]

### id of first order? To get id use 1 rather than 0- 0 must be the identifier for the very first column rather than the first column of data.

In [25]:
baskets.iloc[1,0]

166

In [26]:
baskets.iloc[29304, 2]

'2022-07-31 16:52:33.731'

### Note columns for future reference

In [22]:
baskets.columns

Index(['id', 'order_id', 'placed_at', 'merchant_id', 'sku_id', 'top_cat_id',
       'sub_cat_id', 'qty', 'price'],
      dtype='object')

### Note that the use of numbers to hide product and merchant info may be confusing- look at these numbers as numerical identifiers, not mathematical values.

In [23]:
baskets.sum()

id                                                    4775191340
order_id                                               801118634
placed_at      2021-05-05 11:04:46.5792021-05-06 10:45:02.448...
merchant_id                                             31094175
sku_id                                                  28648236
top_cat_id                                              318204.0
sub_cat_id                                             1488587.0
qty                                                    202840187
price                                          4041550272.440521
dtype: object

In [24]:
baskets.describe()

id      order_id   merchant_id        sku_id    top_cat_id  \
count   29305.000000  29305.000000  29305.000000  29305.000000  29298.000000   
mean   162948.006825  27337.267838   1061.053574    977.588671     10.860946   
std     95745.525000  18450.495064    761.056051    348.157681      8.353926   
min       126.000000     23.000000     10.000000      5.000000      1.000000   
25%     80106.000000  10296.000000    352.000000    755.000000      4.000000   
50%    158748.000000  25824.000000    934.000000    875.000000      9.000000   
75%    242651.000000  43776.000000   1717.000000   1055.000000     15.000000   
max    338874.000000  61439.000000   3160.000000   2383.000000     35.000000   

         sub_cat_id           qty         price  
count  29298.000000  2.930500e+04  2.930500e+04  
mean      50.808485  6.921692e+03  1.379133e+05  
std       31.280838  1.168396e+06  1.355361e+05  
min        1.000000  1.000000e+00  0.000000e+00  
25%       30.000000  1.000000e+00  4.550000e+04  
50%       48.000000  2.000000e+00  1.070000e+05  
75%       79.000000  5.000000e+00  1.845000e+05  
max      108.000000  2.000000e+08  2.175000e+06

### From describe function, we find that the average order is for 6.921 X 10^3 items, and the average price per order is 1.379 X 10^5 dollars.

## Trying to learn covariance meaning and use

In [28]:
baskets.cov()

id      order_id   merchant_id        sku_id  \
id           9.167206e+09  1.761676e+09  2.894300e+07  8.746761e+06   
order_id     1.761676e+09  3.404208e+08  5.477361e+06  1.676017e+06   
merchant_id  2.894300e+07  5.477361e+06  5.792063e+05  3.689526e+04   
sku_id       8.746761e+06  1.676017e+06  3.689526e+04  1.212138e+05   
top_cat_id   1.181186e+04  2.416692e+03  2.024899e+02  2.373937e+02   
sub_cat_id   5.091529e+04  9.535837e+03  3.478746e+02  1.788915e+03   
qty          7.612809e+08  1.542931e+08  1.269334e+07  3.833156e+06   
price        2.785898e+08  4.288260e+07  8.333774e+04  1.470711e+06   

               top_cat_id     sub_cat_id           qty         price  
id           11811.857955   50915.294886  7.612809e+08  2.785898e+08  
order_id      2416.691685    9535.836604  1.542931e+08  4.288260e+07  
merchant_id    202.489916     347.874643  1.269334e+07  8.333774e+04  
sku_id         237.393656    1788.915178  3.833156e+06  1.470711e+06  
top_cat_id      69.788084      47.294869 -4.646834e+04 -2.984864e+03  
sub_cat_id      47.294869     978.490849 -1.351184e+05  1.221353e+05  
qty         -46468.343996 -135118.437323  1.365150e+12 -9.540640e+08  
price        -2984.863562  122135.260434 -9.540640e+08  1.837003e+10

### What does this mean? Should this function be called on less data?

## Parsing out individual rows/ columns:

In [29]:
baskets['id']

0           126
1           166
2           167
3           168
4           169
          ...  
29300    338870
29301    338871
29302    338872
29303    338873
29304    338874
Name: id, Length: 29305, dtype: int64

### Specific row and column: df.loc[row,column] *has to be column name, not number*

In [36]:
baskets.loc[2]

id                                 167
order_id                            42
placed_at      2021-05-06 10:45:04.850
merchant_id                        196
sku_id                             341
top_cat_id                         3.0
sub_cat_id                        47.0
qty                                  2
price                              0.0
Name: 2, dtype: object

In [37]:
baskets.loc[346,'id']

4922

# Start answering some questions

#### 1. How much are the top ten merchants spending per order on average? (Top ten most often buying)
- Start by finding top ten most frequent merchants

In [44]:
list = baskets['merchant_id'].value_counts()[:10].index.tolist()
print(list)

[352, 1753, 951, 496, 349, 185, 1067, 488, 168, 398]


##### - Next, find how many orders each merchant has made total, in the order of the previous list

In [45]:
for i in range(10):
 print(baskets[baskets['merchant_id'] == list[i]].count()['id'])

1986
1163
950
791
765
758
754
677
659
594


#### Now try to add up the cost of one merchant's total orders over time (Use first one for example)

In [46]:
print(baskets[baskets['merchant_id'] == 352].sum()['price'])

332151900.0


### Similar function as before to get all 10 outputs:

In [47]:
for i in range(10):
 print(baskets[baskets['merchant_id'] == list[i]].sum()['price'])

332151900.0
150617400.0
135886476.0
103489034.72727272
85935084.0
73866200.0
105858106.0
99062176.42857143
73859850.0
61358900.0


#### Maximum spent was merchant_id 352 with 332151900.0. Is this the same max as the overall spent? Question to explore later further.

#### Now let's find how much they all spend on average per order by dividing thier purchasing total by the number of purchases made.

In [48]:
for i in range(10):
    print(baskets[baskets['merchant_id'] == list[i]].sum()['price'] / baskets[baskets['merchant_id'] == list[i]].count()['id'])

167246.67673716013
129507.65262252795
143038.3957894737
130833.16653258246
112333.4431372549
97448.81266490766
140395.36604774537
146325.223675881
112078.67981790591
103297.81144781144


#### In conclusion, I found that the most common purchasing merchant is #352, who also spent the most in total out of the most common purchasers, and spends the most on average per order out of the 10 most common purchasers. This shows that #352 is a highly valuable customer, later it may be useful to observe their purchasing patterns over time. 
#### Further questions:
- Is #352 the highest spender over all merchants?
- How many of the most common purchasers are also in the list of highest paying customers?
- How much are the top ten highest paying merchants spending per order?